File: ios/Services/SignetsServiceProtocol.swift
//
//  SignetsServiceProtocol.swift
//  ETSMobile WidgetExtension
//
//  Created by Jonathan Duval-Venne on 2022-11-07.
//

import Foundation

protocol SignetsServiceProtocol {
    /// Call the SignetsAPI to get the courses activities for the [session] for
    /// the student ([username]). By specifying [courseGroup] we can filter the
    /// results to get only the activities for this course.
    /// If the [startDate] and/or [endDate] are specified the results will contains
    /// all the activities between these dates
    func getCoursesActivities(user: MonETSUser, session: String, courseGroup: String, startDate: Date?, endDate: Date?, completion: @escaping (Result<[CourseActivity], ApiError>) -> ());

    /// Call the SignetsAPI to get the courses activities for the [session] for
    /// the student ([username]).
    func getScheduleActivities(user: MonETSUser, session: String, completion: @escaping (Result<[ScheduleActivity], ApiError>) -> ());

    /// Call the SignetsAPI to get the courses of the student ([username]).
    func getCourses(user: MonETSUser, completion: @escaping (Result<[Course], ApiError>) -> ());

    /// Call the SignetsAPI to get all the evaluations (exams) and the summary
    /// of [course] for the student ([username]).
    func getCourseSummary(user: MonETSUser, course: Course, completion: @escaping (Result<CourseSummary, ApiError>) -> ());

    /// Call the SignetsAPI to get the list of all the [Session] for the student ([username]).
    func getSessions(user: MonETSUser, completion: @escaping (Result<[Session], ApiError>) -> ());

    /// Call the SignetsAPI to get the [ProfileStudent] for the student.
    func getStudentInfo(user: MonETSUser, completion: @escaping (Result<ProfileStudent, ApiError>) -> ());

    /// Call the SignetsAPI to get the list of all the [Program] for the student ([username]).
    func getPrograms(user: MonETSUser, completion: @escaping (Result<[Program], ApiError>) -> ());

    /// Call the SignetsAPI to get the list of all [CourseReview] for the [session]
    /// of the student ([username]).
    func getCourseReviews(user: MonETSUser, session: Session, completion: @escaping (Result<[CourseReview], ApiError>) -> ());
}

File: android/app/src/main/kotlin/ca/etsmtl/applets/etsmobile/services/SignetsServiceProtocol.kt
package ca.etsmtl.applets.etsmobile.services

import ca.etsmtl.applets.etsmobile.services.models.MonETSUser
import ca.etsmtl.applets.etsmobile.services.models.Semester

interface SignetsServiceProtocol {
    fun getSessions(user: MonETSUser, completion: (Result<List<Semester>>) -> Unit)
//    fun getStudentInfo(user: ca.etsmtl.applets.etsmobile.services.models.MonETSUser, completion: (Result<ProfileStudent, ApiError>) -> Unit)
//    fun getCourses(user: ca.etsmtl.applets.etsmobile.services.models.MonETSUser, completion: (Result<List<Course>, ApiError>) -> Unit)
//    fun getCoursesActivities(user: ca.etsmtl.applets.etsmobile.services.models.MonETSUser, session: String, courseGroup: String, startDate: Date?, endDate: Date?, completion: (Result<List<CourseActivity>, ApiError>) -> Unit)
//    fun getScheduleActivities(user: ca.etsmtl.applets.etsmobile.services.models.MonETSUser, session: String, completion: (Result<List<ScheduleActivity>, ApiError>) -> Unit)
//    fun getCourseSummary(user: ca.etsmtl.applets.etsmobile.services.models.MonETSUser, course: Course, completion: (Result<CourseSummary, ApiError>) -> Unit)
//    fun getPrograms(user: ca.etsmtl.applets.etsmobile.services.models.MonETSUser, completion: (Result<List<Program>, ApiError>) -> Unit)
//    fun getCourseReviews(user: ca.etsmtl.applets.etsmobile.services.models.MonETSUser, session: ca.etsmtl.applets.etsmobile.services.models.Session, completion: (Result<List<CourseReview>, ApiError>) -> Unit)
}
